tests/DCPS/DynamicDataWriter/publisher.cpp

```diff
@@ -0,0 +1,85 @@
+#include <dds/DCPS/StaticIncludes.h>
+#include <dds/DCPS/DCPS_Utils.h>
+#include <dds/DCPS/XTypes/DynamicTypeSupport.h>
+#include <dds/DCPS/XTypes/DynamicDataImpl.h>
+#ifdef ACE_AS_STATIC_LIBS
+#  include <dds/DCPS/RTPS/RtpsDiscovery.h>
+#  include <dds/DCPS/transport/rtps_udp/RtpsUdp.h>
+#endif
+
+bool check_rc(DDS::ReturnCode_t rc, const char* what)
+{
+  if (rc != DDS::RETCODE_OK) {
+    ACE_ERROR((LM_ERROR, "publisher (%P|%t) ERROR: %C: %C\n",
+      what, OpenDDS::DCPS::retcode_to_string(rc)));
+    return true;
+  }
+  return false;
+}
+
+int ACE_TMAIN(int argc, ACE_TCHAR *argv[])
+{
+  // Production apps should check the return values.
+  // For tests, it just makes the code noisy.
+
+  // Coordination across processes.
+  DistributedConditionSet_rch distributed_condition_set =
+    OpenDDS::DCPS::make_rch<FileBasedDistributedConditionSet>();
+
+  DDS::DomainParticipantFactory_var domain_participant_factory = TheParticipantFactoryWithArgs(argc, argv);
+  DDS::DomainParticipant_var participant =
+    domain_participant_factory->create_participant(HelloWorld::HELLO_WORLD_DOMAIN,
+                                                   PARTICIPANT_QOS_DEFAULT,
+                                                   0,
+                                                   0);
+
+  HelloWorld::MessageTypeSupport_var type_support = new HelloWorld::MessageTypeSupportImpl();
+  CORBA::String_var type_name = type_support->get_type_name();
+
+  DDS::DynamicType_var dt = type_support->get_type();
+  DDS::TypeSupport_var dts = new DDS::DynamicTypeSupport(dt);
+  if (check_rc(dts->register_type(participant, type_name), "register_type failed")) {
+    return 1;
+  }
+
+  DDS::Topic_var topic = participant->create_topic(HelloWorld::MESSAGE_TOPIC_NAME,
+                                                   type_name,
+                                                   TOPIC_QOS_DEFAULT,
+                                                   0,
+                                                   0);
+
+  DDS::Publisher_var publisher = participant->create_publisher(PUBLISHER_QOS_DEFAULT,
+                                                               0,
+                                                               0);
+
+  DDS::DataWriter_var data_writer = publisher->create_datawriter(topic,
+                                                                 DATAWRITER_QOS_DEFAULT,
+                                                                 0,
+                                                                 0);
+  DDS::DynamicDataWriter_var ddw = DDS::DynamicDataWriter::_narrow(data_writer);
+
+  distributed_condition_set->wait_for(HelloWorld::PUBLISHER, HelloWorld::SUBSCRIBER, HelloWorld::SUBSCRIBER_READY);
+
+  DDS::DynamicData_var dd = new OpenDDS::XTypes::DynamicDataImpl(dt);
+  if (check_rc(dd->set_string_value(0, "Hello, World!"), "set_string_value failed")) {
+    return 1;
+  }
+
+  if (check_rc(ddw->write(dd, DDS::HANDLE_NIL), "write failed")) {
```

Comment:
Looks like `write` takes ownership of the pointer in `dd` and causes a memory issue when the destructor of `dd` is invoked.

Reply:
Specifically the constructors of `DynamicSample` contain ptr-to-var initialization so those need `_duplicate`.